packages/headless/src/controllers/commerce/recommendations/headless-recommendations.ts

```diff
@@ -0,0 +1,118 @@
+import {Schema} from '@coveo/bueno';
+import {createSelector} from '@reduxjs/toolkit';
+import {CommerceAPIErrorStatusResponse} from '../../../api/commerce/commerce-api-error-response';
+import {Product} from '../../../api/commerce/common/product';
+import {
+  CommerceEngine,
+  CommerceEngineState,
+} from '../../../app/commerce-engine/commerce-engine';
+import {configuration} from '../../../app/common-reducers';
+import {contextReducer as commerceContext} from '../../../features/commerce/context/context-slice';
+import {recommendationSlotDefinition} from '../../../features/commerce/recommendation/recommendation';
+import {
+  fetchRecommendations,
+  updateRecommendationSlotId,
+} from '../../../features/commerce/recommendation/recommendation-actions';
+import {recommendationV2Reducer as recommendation} from '../../../features/commerce/recommendation/recommendation-slice';
+import {loadReducerError} from '../../../utils/errors';
+import {validateInitialState} from '../../../utils/validate-payload';
+import {
+  buildController,
+  Controller,
+} from '../../controller/headless-controller';
+
+/**
+ * The `Recommendations` controller exposes a method for retrieving recomendations content in a commerce interface.
+ */
+export interface Recommendation extends Controller {
+  /**
+   * Fetches the recommendations.
+   */
+  refresh(): void;
+
+  /**
+   * A scoped and simplified part of the headless state that is relevant to the `Recommendation` controller.
+   */
+  state: RecommendationState;
+}
+
+export interface RecommendationState {
+  headline: string;
+  products: Product[];
+  error: CommerceAPIErrorStatusResponse | null;
+  isLoading: boolean;
+  responseId: string;
+}
+
+export type RecommendationControllerState = Recommendation['state'];
+
+interface RecommendationsOptions {
+  slotId: string;
+}
+
+interface RecommendationsProps {
+  options: RecommendationsOptions;
+}
+
+function validateRecommendationInitialState(
+  engine: CommerceEngine,
+  state: RecommendationsOptions
+) {
+  if (!state) {
+    return;
+  }
+
+  const schema = new Schema<RecommendationsOptions>({
+    slotId: recommendationSlotDefinition,
+  });
```

Comment:
This schema definition can be moved to [features/commerce/recommendation/recommendation.ts](https://github.com/coveo/ui-kit/pull/3734/files#diff-6f834abc4da1172ac499c73f2e9334b33f60943e27dafc68155da2b018362ff5) and reused in the payload validation in `updateRecommendationsSlotId`: [features/commerce/recommendation/recommendation-actions.ts](https://github.com/coveo/ui-kit/pull/3734/files#diff-1a9264e1774a126d8783aa61a25d80bf628f26b77acdcf58dbdd40cb4cf21b5c)